### ISSUE TITLE
Add a way to access files outside of the App directory

### ISSUE THREAD

**Issue by jasells** [praeclarum/sqlite-net#851]
Background: UWP requires user permission to write in directories outside of the app's sandbox. SQLIte by default opens read+write. This post addresses the root of the [problem](https://stackoverflow.com/questions/35084058/uwp-app-sqlite-access-database-in-documents-library) very well, so I won't repeat.

I'll take a look to offer a better implementation suggestion, but it seems like there could be a fairly simple way to ask for permission (show a picker) on UWP when creating a connection (from a path).  

Another option could be to allow open in read only (via connection constructor)?  This, would of course result in errors If SQL commands result in saves....  but there is a large use case, I think for allowing a read-only connection mode.  

My specific case is that I have another library I'm using that has large (perhaps Gigabytes) data stored in a SQLite file, and I need to be able to allow the user to load these data sources in my UWP app that may be user created by other app(s).

**Comment by fnc12** [comment]
I am facing the same issue. Everything looks well but one cannot access files outside the app dir even if my app has restricted permission `broadFileSystemAccess` in its manifest and file system access is allowed for my app inside Win10 Settings app. This makes SQLite libs useless for UWP (including this lib) if user want to open a large SQLite file - I don't want to copy every database file inside app dir cause this is a bad behavior. 
BTW this lib is great

**Comment by jasells** [comment]
The issue on UWP is write access required.  To get write access, even with the manifest declarations, the user still needs to give permission at runtime.  Typically, this can be done with the file picker. Sent from Mail for Windows 10 From: Yevgeniy ZakharovSent: Monday, March 9, 2020 8:30 AMTo: praeclarum/sqlite-netCc: jasells; AuthorSubject: Re: [praeclarum/sqlite-net] Add a way to access files outside of the App directory (#851) I am facing the same issue. Everything looks well but one cannot access files outside the app dir even if my app has restricted permission broadFileSystemAccess in its manifest and file system access is allowed for my app inside Win10 Settings app. This makes SQLite libs useless for UWP (including this lib) if user want to open a large SQLite file - I don't want to copy every database file inside app dir cause this is a bad behavior.BTW this lib is great—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by jasells** [comment]
Just as an observation....   seems like everyone was all over MS for lack of security 10 years ago, now, their the most secure, but ppl won't adjust their code to support...

**Comment by jasells** [comment]
Just FYI, removable storage, like SD cards or USB drives, are not as restricted in UWP.  My best current work around is to put the sqllit file there.

**Comment by fnc12** [comment]
what do you mean `this can be done with the file picker`? I use picker:
```c#
            var picker = new Windows.Storage.Pickers.FileOpenPicker
            {
                ViewMode = Windows.Storage.Pickers.PickerViewMode.Thumbnail,
                SuggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.DocumentsLibrary
            };
            Windows.Storage.StorageFile file = await picker.PickSingleFileAsync();
            if (file != null)
            {
                var path = file.Path;
                //  do something with file at path
            }
```
but still facing this issue.

**Comment by jasells** [comment]
> what do you mean `this can be done with the file picker`? I use picker:
> 
> ```cs
> var picker = new Windows.Storage.Pickers.FileOpenPicker
> {
>     ViewMode = Windows.Storage.Pickers.PickerViewMode.Thumbnail,
>     SuggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.DocumentsLibrary
> };
> Windows.Storage.StorageFile file = await picker.PickSingleFileAsync();
> if (file != null)
> {
>     var path = file.Path;
>     //  do something with file at path
> }
> ```
> 
> but still facing this issue.

You are right.  The file picker can't solve it, because SqlLite API doesn't support the provided filestream as input.  Been a while since I posted original.  =/

See above comment for my current workaround (save the file on removable storage).

**Comment by fnc12** [comment]
I use PC. I don’t have a removable storage. I have drive C: and that is all

Отправлено с iPhone

**Comment by jasells** [comment]
> I use PC. I don’t have a removable storage. I have drive C: and that is all Отправлено с iPhone
> […](#)
> 10 марта 2020 г., в 18:23, jasells ***@***.***> написал(а): what do you mean this can be done with the file picker? I use picker: var picker = new Windows.Storage.Pickers.FileOpenPicker { ViewMode = Windows.Storage.Pickers.PickerViewMode.Thumbnail, SuggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.DocumentsLibrary }; Windows.Storage.StorageFile file = await picker.PickSingleFileAsync(); if (file != null) { var path = file.Path; // do something with file at path } but still facing this issue. You are right. The file picker can't solve it, because SqlLite API doesn't support the provided filestream as input. Been a while since I posted original. =/ See above comment for my current workaround (save the file on removable storage). — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

A usb-flash drive is removable storage.  Again, not a great solution, but it the best I've come up with sans SqlLite adding a stream input to the API, or internally requesting the File picker to have the user select the file.  (I *think* after the user selects a *particular* file once, the app can access that file again without requesting *every* time).
